query_merge.go

```diff
@@ -0,0 +1,296 @@
+		return fn(q)
+	}
+	return q
+}
+
+func (q *MergeQuery) With(name string, query schema.QueryAppender) *MergeQuery {
+	q.addWith(name, query, false)
+	return q
+}
+
+func (q *MergeQuery) WithRecursive(name string, query schema.QueryAppender) *MergeQuery {
+	q.addWith(name, query, true)
+	return q
+}
+
+//------------------------------------------------------------------------------
+
+func (q *MergeQuery) Table(tables ...string) *MergeQuery {
+	for _, table := range tables {
+		q.addTable(schema.UnsafeIdent(table))
+	}
+	return q
+}
+
+func (q *MergeQuery) TableExpr(query string, args ...interface{}) *MergeQuery {
+	q.addTable(schema.SafeQuery(query, args))
+	return q
+}
+
+func (q *MergeQuery) ModelTableExpr(query string, args ...interface{}) *MergeQuery {
+	q.modelTableName = schema.SafeQuery(query, args)
+	return q
+}
+
+//------------------------------------------------------------------------------
+
+// Returning adds a RETURNING clause to the query.
+//
+// To suppress the auto-generated RETURNING clause, use `Returning("NULL")`.
+// Only for mssql output, postgres not supported returning in merge query
+func (q *MergeQuery) Returning(query string, args ...interface{}) *MergeQuery {
+	q.addReturning(schema.SafeQuery(query, args))
+	return q
+}
+
+//------------------------------------------------------------------------------
+
+func (q *MergeQuery) Operation() string {
+	return "MERGE"
+}
+
+func (q *MergeQuery) AppendQuery(fmter schema.Formatter, b []byte) (_ []byte, err error) {
+	if q.err != nil {
+		return nil, q.err
+	}
+
+	fmter = formatterWithModel(fmter, q)
+
+	b, err = q.appendWith(fmter, b)
+	if err != nil {
+		return nil, err
+	}
+
+	b = append(b, "MERGE "...)
+	if q.db.dialect.Name() == dialect.PG {
+		b = append(b, "INTO "...)
+	}
+
+	b, err = q.appendFirstTableWithAlias(fmter, b)
+	if err != nil {
+		return nil, err
+	}
+
+	b = append(b, " USING "...)
+	b, err = q.using.AppendQuery(fmter, b)
+	if err != nil {
+		return nil, err
+	}
+
+	b = append(b, " ON "...)
+	b, err = q.on.AppendQuery(fmter, b)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, w := range q.when {
+		b = append(b, " WHEN "...)
+		b, err = w.AppendQuery(fmter, b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if q.hasFeature(feature.Output) && q.hasReturning() {
+		b = append(b, " OUTPUT "...)
+		b, err = q.appendOutput(fmter, b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	//A MERGE statement must be terminated by a semi-colon (;).
+	b = append(b, ";"...)
+
+	return b, nil
+}
+
+//------------------------------------------------------------------------------
+
+func (q *MergeQuery) Using(s string, args ...interface{}) *MergeQuery {
+	q.using = schema.SafeQuery(s, args)
+	return q
+}
+
+func (q *MergeQuery) On(s string, args ...interface{}) *MergeQuery {
+	q.on = schema.SafeQuery(s, args)
+	return q
+}
+
+// WhenUpdate for when update clause
+func (q *MergeQuery) WhenUpdate(expr string, fn func(q *UpdateQuery) *UpdateQuery) *MergeQuery {
```

Comment:
Perhaps we can also accept `args ...interface{}` as the last argument? It is a bit awkward, but would allow something like:

```go
q.When("MATCHED AND foo = ?", func() {...}, 123)
```